source/server/server.cc

```diff
@@ -141,14 +143,22 @@ void InstanceImpl::flushStats() {
   // A shutdown initiated before this callback may prevent this from being called as per
   // the semantics documented in ThreadLocal's runOnAllThreads method.
   stats_store_.mergeHistograms([this]() -> void {
-    HotRestart::GetParentStatsInfo info;
-    restarter_.getParentStats(info);
+    uint64_t parent_memory_allocated = 0;
+    uint64_t parent_connections = 0;
+    std::unique_ptr<HotRestartMessage> wrapper_msg = restarter_.getParentStats();
```

Comment:
Don't we need to store this somewhere so that when someone asks for stats via a `/stats` admin endpoint we provide pull based stats for both processes?

Reply:
Sure, should be straightforward. Just to make sure I'm doing the right thing: you mean that `/stats` should show both the parent stats, and the combined parent+child stats? And, since that was not possible when the combination was handled by writing to shared memory, this would entail adding to what is displayed by `/stats`?

(In case you're talking about the server_stats_ parent-specific ones, those values are still recorded as their own thing. Their values were previously passed in the GetParentStatsInfo struct, are still passed and recorded.

Reply:
No, this is keeping the behavior consistent. Because of shared memory, `/stats` always showed both the old process and the new process. We have to keep that behavior, so we should save the parent data during every flush interval and just use it for computing `/stats`. This is related to your other PR that got reverted. It might be worth it to fix that first + tests so you get some familiarity with that path, and this will make more sense.

Reply:
By "showed both the old process and the new process" you mean showing a single copy of each stat, whose value would be the same for both old and new, right? Since, due to use of shared memory, they couldn't be anything but identical. That's what I gather from looking at AdminImpl::handlerStats(): it's just server_.stats(), which was shared memory until now.

The code in this PR behaves exactly the same: it presents the combined child+parent values, because that is what the child holds in server_.stats(). The combination is done as soon as the stats update RPC arrives, and the child stores the result as its actual stats values. In particular, I am *not* keeping the child+parent values separate, and combining them just when it's time to present them to the outside world.

>This is related to your other PR that got reverted.

My understanding was that the BoolIndicator revert was just because the stats exporters weren't reaching into store.boolIndicators(), and one of the converted-over gauges was important enough to someone that is was best to just revert ASAP, rather than add code to export it. I think that fix can proceed independently.

Reply:
> The code in this PR behaves exactly the same: it presents the combined child+parent values, because that is what the child holds in server_.stats(). The combination is done as soon as the stats update RPC arrives, and the child stores the result as its actual stats values. In particular, I am not keeping the child+parent values separate, and combining them just when it's time to present them to the outside world.

Yup OK sorry, I didn't look at things in enough detail, this should be fine. I will review the whole thing in more detail once we resolve the merge conflicts.

> I think that fix can proceed independently.

I think this PR depends on that PR right? Probably easier to fix that one first and resubmit? Up to you.

Reply:
Ah, correct: BoolIndicator was meant to pave the way for this PR. However, when I considered working on this PR while that one is rolled back, it didn't seem like much of a hurdle. So, now I'm actually not sure if it's worth bringing back. The "stat combination logic exception" code is looking to be reasonable enough that maybe adding ~10 extra exception entries is actually cleaner than adding all the BoolIndicator machinery.

Reply:
How would you manage the 10 exceptions? Would that scale over time? Are any of the exceptions per-cluster?

The concept of a non-arithmetic stat makes a lot of sense to me as a separate class. But I think I mentioned this a while back you could use one of the flag bits for that if subclassing or alternate-classing is too much work.

Reply:
It would look like additional {"statname", CombineLogic::SomeLogic} items in combine_logic_exceptions over in HotRestartingChild. Computational complexity-wise, it would be ok - either there would be too few to care, or we could go to a trie. Code complexity-wise, it would just be a large-ish collections of constants. So, maybe pretty ugly by like 50 bools.

As for the flags approach, I think it would not have an impact on the type of complexity that is the issue here, i.e. the BoolIndicator PR's "47 files changed." Fundamentally, the choice is to build this distinction into the stats system itself as a new type with its own default logic, or to specify some additional exceptions for the existing gauge type. The fact that a new type would need support throughout the whole stats machine is what makes it a real tradeoff.

Ignoring all of these issues, the non-arithmetic stat idea certainly does make a lot of conceptual sense to me as well :)

Reply:
My preference in general is to keep this as simple as possible until we decide we need something more complicated. @fredlas what would be the simplest solution that would cover our current needs? Does this PR need changes to bring us that solution, or do we need the other PR back? Basically I'm just looking for a summary before digging back in. :)

Reply:
Thanks for putting it that way - that makes it straightforward. The right approach from that perspective is definitely "proceed without BoolIndicator." This PR will just drop the little bit of indicator-specific machinery that the value combination logic previously had, and add entries to the gauge value combination logic exceptions map for what would have been bools.